clusterloader2/pkg/measurement/common/pod_command.go

```diff
@@ -0,0 +1,697 @@
+func (p *podPeriodicCommandMeasurement) getMaxNPods(n int) []*v1.Pod {
+	store := p.informer.GetStore()
+	pods := []*v1.Pod{}
+
+	podList := store.List()
+	if len(podList) == 0 {
+		return pods
+	}
+
+	for _, podInterface := range podList {
+		pod := podInterface.(*v1.Pod)
+		if !p.isApplicablePod(pod) {
+			continue
+		}
+
+		pods = append(pods, pod)
+
+		if len(pods) >= n {
+			return pods
+		}
+	}
+
+	return pods
+}
+
+// runCommandInPod runs a specific given command in the specific given pod.
+func (p *podPeriodicCommandMeasurement) runCommandInPod(
+	pod *v1.Pod, params *podPeriodicCommandMeasurementCommandParams,
+) (*runCommandResult, error) {
+	klog.V(4).Infof(
+		"%s: running named command %s in pod %s/%s",
+		podPeriodicCommandMeasurementName, params.Name, pod.Namespace, pod.Name,
+	)
+
+	p.statsLock.Lock()
+	p.stats.Execs++
+	p.statsLock.Unlock()
+
+	req := p.clientset.CoreV1().RESTClient().
+		Post().
+		Namespace(pod.Namespace).
+		Resource("pods").
+		Name(pod.Name).
+		SubResource("exec").
+		VersionedParams(&v1.PodExecOptions{
+			Container: p.params.Container,
+			Command:   params.Command,
+			Stdin:     false,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	executor, err := remotecommand.NewSPDYExecutor(p.restConfig, "POST", req.URL())
+	if err != nil {
+		return nil, err
+	}
+
+	exitCode := 0
+	stdoutBuf := &bytes.Buffer{}
+	stderrBuf := &bytes.Buffer{}
+	hitTimeout := false
+	// Holds error returned from executor.Stream.
+	execErrChan := make(chan error, 1)
+	// This value is set below when the command finishes or when we hit a timeout.
+	var endTime time.Time
+
+	// The logic used here to start the executor and the timeout timer isn't super precise, but
+	// it is good enough for this use case. It is ok that the timeout timer is started after the
+	// executor, since we still guarantee that the timeout is at least the configured value.
+	startTime := time.Now()
+
+	go func() {
+		err := executor.Stream(remotecommand.StreamOptions{
+			Stdout: stdoutBuf,
+			Stderr: stderrBuf,
+		})
+		execErrChan <- err
+	}()
+
+	t := time.NewTimer(params.Timeout)
+	defer t.Stop()
+
+	// Two different cases: (1) if the command returns before the timeout, and (2) if the timeout
+	// triggers before the command is done.
+	// If the timeout triggers, then the command isn't cancelled. This logic isn't available until
+	// client-go version 0.26 (see Executor.StreamWithContext).
+	select {
+	case err = <-execErrChan:
+		endTime = time.Now()
+		if err == nil {
+			break
+		}
+
+		switch e := err.(type) {
+		case exec.CodeExitError:
+			p.statsLock.Lock()
+			p.stats.NonZeroRCs++
+			p.statsLock.Unlock()
+
+			exitCode = e.ExitStatus()
+
+			klog.V(2).Infof(
+				"%s: warning: non-zero exit code %d for named command %s in pod %s/%s",
+				podPeriodicCommandMeasurementName, exitCode, params.Name, pod.Namespace, pod.Name,
+			)
+		default:
+			return nil, err
+		}
+	case <-t.C:
```

Comment:
We could simplify it a bit with `time.After(timeout)` and get rid of timer.

Reply:
I'm not sure if we could. This code checks if a timeout occurs, but it is also structured so that the function can return early when this happens. If we use `time.After(timeout)` and remove the timer, then I don't think we would be able to return as soon as a timeout occurs. I think we could only check if a command took longer than the timeout after the command has finished on its own.